### ISSUE TITLE
Pyright reports incorrect return types when aliasing Callable in class decorator types starting in 1.1.370

### ISSUE THREAD

**Issue by ryanbagwell** [microsoft/pyright#9145]
**Describe the bug**

### Consider this typical decorator pattern:

```
def generic_decorator(
    name: str,
) -> Callable[
    [Callable[P, T]],
   Callable[P, T]
]:

    def inner(func_or_cls: Callable[P, T]) -> Callable[P, T]:

        @wraps(func_or_cls)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
        
            return func_or_cls(*args, **kwargs)
        
        return wrapper

    return inner

@generic_decorator(name="Steve")
class GenericClass:
    def a_method(self) -> None:
        pass

GenericClass().a_method()
```

Pyright correctly reports the return type of `a_method()` is `None` and suggests properties on the `GenericClass()` instance:

<img width="600" alt="Screenshot 2024-10-03 at 5 13 56 PM" src="https://github.com/user-attachments/assets/ddc3d1a6-f400-49fd-982e-f4818944224e">

<img width="597" alt="Screenshot 2024-10-03 at 5 13 36 PM" src="https://github.com/user-attachments/assets/88db414a-bcd9-41e2-9c46-0434972b02d6">

### However, if we alias `[Callable[P, T]` and use that alias, the return type of the instance is unknown:

```
MethodOrClass: TypeAlias = Callable[P, T]

def alias_decorator(
    name: str,
) -> Callable[
    [MethodOrClass],
    MethodOrClass
]:

    def inner(func_or_cls: MethodOrClass) -> MethodOrClass:

        @wraps(func_or_cls)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> MethodOrClass:
        
            return func_or_cls(*args, **kwargs)
        
        return wrapper

    return inner

@alias_decorator(name="Steve")
class AliasClass:
    def a_method(self) -> None:
        pass

AliasClass().a_method()
```

Pyright loses the return type of `a_method() does not suggest properties on the `AliasClass()` instance.

<img width="244" alt="Screenshot 2024-10-03 at 5 15 48 PM" src="https://github.com/user-attachments/assets/130f1bf6-49ed-438b-82a0-f26eda4d5da0">

<img width="338" alt="Screenshot 2024-10-03 at 5 15 37 PM" src="https://github.com/user-attachments/assets/cf6dc340-f158-4791-a937-ec81f43c2e5b">

This worked in pyright (via pylance) up to v1.1.369, but broke in v1.1.370:

[Reproducible in this playground](https://pyright-play.net/?pyrightVersion=1.1.370&pythonVersion=3.10&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDRQAKRIREAygtQMb0Aqi1AIIkkRAM58BANWYAoUJCjAArii4wwYEmMzZc%2BAO4sEY2bN5QAvFH6cZIABQAiXk4CUshlcbNWHbs4M7mZOoVDmAgAWqDA6ymKoGGjUKNQgSFyEpORU1FAGuADWmChQAIwAdJUAzABsAJxQRCgAJk0UYABu1LKhTmYt1MBQyanpXAD6g1y4RBqOslBLUCis1ABcUGIwILSyblAAtAB8WWSUNADai8uXxOe5lwx8ALove8tnOVfPNi%2ByL3WZk%2Bg2GqDGDhUagmuAmXG0m3u32oT1eBxOXwuKN%2BvEBwM%2BywAAkYiCZIapJrD4WIPATlqD8sZOI4AFTMNBiTYMCrsiRQFkswoGXlcipC3no068IF0pY3WXLEDUGDKEClKGUkBw7QONkgDn0AXi-U0%2BUEs2fJUqtWM0nM-FLK2q0rgtJmQmjNIZKbcWbzByrCDUSxONgwajdYLw8Q6ADiKS9XAI5DEnLNDKIEyDMEiYBaDjE1BIwElUAAcmBUjLZQgY2YAMRQACSpUqNQA7AAGeg5pA6MS55QkNpiZRoZLbWCROZNLPK3NtPtQStQYSiMTJmMONz0ZqL-AtMDUUyNgASXTSy%2B6ICneWzC6arUw%2BDERDgOicDnveYOGbnObzbcjlOCtUn6eMxgyTdU23Hl-wXbczAgxNoLEWCQjCWRzwMCM0noJBhhwpoRHETFHhxF5HzaeI8hzGdSHXEptmoIgWnwwwih0VBZDbCo6kaChlHwFAwHwVAoFOXjqi7XpQjMABZec8wAeRAVDNlsIQSJ0awkSxVE-gGIZiPXH0ZhYf0zUDDYth2PZSz0x4zUuRSAJaVTUPeKBXIXDyU1MPF02M11HA1GEtWpTYfJUtT-NLaL3NimMgQtIkSTJMKqW0WlZQZdLmV1EVGB5E1DUFYUTVFY0OXipTEvU1K5QVS1lWdJQKXC7U0L1A1%2BXKiVGvCBUnRtfK3TNEaXRQMZ3QY8QzL9XAAzWEMw1wqN-NXbSGpBYzM2-fNC2LUtQI2Qba1TBtm1bKo%2BK7ehc2vRIr0vHM7zqrY3xPKByTUGAkErA59rqzYAFUUEKESDBQWQ13EVDYOBtzENh7b-PQoA)

**Comment by erictraut** [comment]
Pyright is working correctly here.

You've defined a generic type alias that accepts two type arguments — a ParamSpec and a TypeVar. Wherever you use this type alias, you should be providing two type arguments, but you're omitting these. That means a type checker will assume default values for these type arguments. In the case of a ParamSpec, the default value is `...`, and in the case of a TypeVar the default value is `Any`. I don't think that's what you're intending.

Here's what I think you intended to do:

```python
def alias_decorator(
    name: str,
) -> Callable[[MethodOrClass[P, T]], MethodOrClass[P, T]]:
    def inner(func_or_cls: MethodOrClass[P, T]) -> MethodOrClass[P, T]:
        @wraps(func_or_cls)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
            return func_or_cls(*args, **kwargs)

        return wrapper

    return inner
```

In your `generic_decorator` example, `P` and `T` are scoped to the `generic_decorator` function and referenced by inner scopes. When these inner scopes refer to `P` and `T`, then are referring to the outer-scoped type variables. Contrast that to your `alias_decorator` example where `P` and `T` are not scoped to the outer function but instead are scoped to the inner `wrapper` function.

Here's a tip that might help you out. If you hover over a type variable in pyright or pylance, you'll see the scope that it's bound to. For example, hovering over `T`, you might see `T@alias_decorator`, which indicates that `T` is bound to the `alias_decorator` function scope.

[Scoping rules for type variables](https://typing.readthedocs.io/en/latest/spec/generics.html#scoping-rules-for-type-variables) are confusing in Python. Python 3.12 introduces [new syntax](https://peps.python.org/pep-0695/) that helps alleviate this confusion. If your code doesn't need to run on version of Python older than 3.12, I recommend switching to the new syntax. This also eliminates the need to allocate TypeVars and ParamSpecs at the global level and then rely on subtle, implicit rules to bind them to scopes. Here's how your code looks if you use the new syntax.

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBMCuB2BjALmMAbAzlAlhADmCMlAO4gCG%2BmAUKJFMgJ747wDmuBRJAwhenQUARugCmNGs3xioAWTHIAFmAAmAeRC8hmTAG0AVAYAKAGigAVALpQAvFH6CR4vWctXJqscCgCcFTAB9L0QiClQQQxNzawAKGihEqHgKCDEALihMZBBTGgBKKABaAD4HASFRMT09BWU1TW0A-TdrK3M6lQ0tHRaYqyt0hKSvHzZ4MRBYuCRAokDELEzOhp7m137C0vlFLsbejfchpJOoAAFyKkxphEQ5kAWsfOHTqFGySnwZKYMKEHZMJljAA6P4A8xGADWpDBgKgIOhsK2ZQsx1epxAilgIHgMFu90e11%2B-0wEIMiJJzxeGKxOI%2BVG%2BkhOmOQ2Nx40mkjOfgCwTEoUoEViKTStgARABlZBiABuYjFz0WzSgAEF0P5ME1dGjEu8KIE0vVVLFMGJ0MBkVAAHJgCY6074ZqSNUarXXfKgg27NSxZ5AA)

```python
from functools import wraps
from typing import Callable

type MethodOrClass[**P, T] = Callable[P, T]

def alias_decorator[**P, T](
    name: str,
) -> Callable[[MethodOrClass[P, T]], MethodOrClass[P, T]]:
    def inner(func_or_cls: MethodOrClass[P, T]) -> MethodOrClass[P, T]:
        @wraps(func_or_cls)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
            return func_or_cls(*args, **kwargs)

        return wrapper

    return inner

@alias_decorator(name="Steve")
class AliasClass:
    def a_method(self) -> None:
        pass

AliasClass().a_method()
```

If the new syntax isn't an option, you may want to at least enable the `reportMissingTypeArgument` check in pyright. This will help you identify cases where you unintentionally omit type arguments for generic classes and type aliases.

**Comment by ryanbagwell** [comment]
@erictraut many thanks for the explanation. I think I made a mistake in my example code. I shouldn't have included the types in the wrapper function arguments (i.e. the `P.args` and P.kwargs` in the `def wrapper(*args: P.args, **kwargs: P.kwargs)`.

If I remove them from [the playground (which I've done)](https://pyright-play.net/?pyrightVersion=1.1.370&pythonVersion=3.10&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDRQAKRIREAygtQMb0Aqi1AIIkkRAM58BANWYAoUJCjAArii4wwYEmMzZc%2BAO4sEY2bN5QAvFH6cZIABQAiXk4CUshlcbNWHbs4M7mZOoVDmAgAWqDA6ymKoGGjUKNQgSFyEpORU1FAGuADWmChQAIwAdJUAzABsAJxQRCgAJk0UYABu1LKhTmYt1MBQyanpXAD6g1y4RBqOslBLUCis1ABcUGIwILSyblAAtAB8WWSUNADai8uXxOe5lwx8ALove8tnOVfPNi%2ByL3WZk%2Bg2GqDGDhUagmuAmXG0m3u32oT1eBxOXwuKN%2BvEBwM%2BywAAkYiCZIapJrD4WIPATlqD8sZOI4AFTMNBiTYMCrsiRQFkswoGXlcipC3no068IF0pY3WXLEDUGDKEClKGUkBw7QONkgDn0AXi-U0%2BUEs2fJUqtWM0nM-FLK2q0rgtJmQmjNIZKbcWbzByrCDUSxONgwajdYLw8Q6ADiKS9XAI5DEnLNDKIEyDMEiYBaDjE1BIwElUAAcmBUjLZQgY2YAMRQACSpUqNQA7AAGeg5pA6MS55QkNpiZRoZLbWCROZNLPK3NtPtQStQYSiMTJmMONz0ZqL-AtMDUUyNgASXTSy%2B6ICneWzC6arUw%2BDERDgOicDnveYOGbnObzbcjlOCtUn6eMxgyTdU23Hl-wXbczAgxNoLEWCQjCWRzwMCM0noJBhhwpoRHETFHhxF5HzaeI8hzGdSHXEptmoIgWnwwwih0VBZDbCo6kaChlHwFAwHwVAoFOXjqi7XpQjMABZec8wAeRAVDNlsIQSJ0awkSxVE-gGIZiPXH0ZhYf0zUDDYth2PZSz0x4zUuRSAJaVTUPeKBXIXDyU1MPF02M11HA1GEtWpTYfJUtT-NLaL3NimMgQtIkSTJMKqW0WlZQZdLmV1XlDUFYUTXipTEvU1K5QVS1lWdJQKXC7U0L1A1%2BRKiVqvCBUnRtfK3TNPqXRQMZ3QY8QzL9XAAzWEMw1wqN-NXbSqpBYzM2-fNC2LUtQI2bra1TBtm1bKo%2BK7ehc2vRIr0vHM7wqrY3xPKByTUGAkErA5NoqzYAFUUEKESDBQWQ13EVDYN%2BtzEPB1b-PQoA), pyright does't complain, but it still doesn't suggest methods on the return type.

But I think what you're saying is that it doesn't really matter because if you don't specify params for the TypeAlias, the TypeChecker will assume the return type is `Any` and pylance won't know which object it should suggest properties for? Do I have that correct?

The idea that you still have to pass arguments and return types to a `TypeAlias` surprises me because it seems like it defeats the purpose of creating a `TypeAlias`.

We're a bit confused because this seems to have changed between between `v1.1.369` and `v1.1.370`. In addition, this isn't a problem for us using the latest version of pycharm.

Also note that we're on Python 3.10.
